```
Waiting up to 3 seconds for cluster initialization.
  
  This will continually check to see if the API for kubernetes is reachable.
  This may time out if there was some uncaught error during start up.

..Cluster failed to initialize within 3 seconds.
Preserving master logs in '/tmp/tmp.zRkaD25gbX'...
Copying e2e-test-mml-master:/var/log/startupscript.log
Warning: Permanently added '130.211.127.7' (ECDSA) to the list of known hosts.
startupscript.log                                                                                                                                                    100%   36KB  35.8KB/s   00:00
Copying e2e-test-mml-master:/var/log/kube-apiserver.log
Warning: Permanently added '130.211.127.7' (ECDSA) to the list of known hosts.
scp: /var/log/kube-apiserver.log: No such file or directory
ERROR: (gcloud.compute.copy-files) [/usr/bin/scp] exited with return code [1]. See https://cloud.google.com/compute/docs/troubleshooting#ssherrors for troubleshooting hints.
2016/02/12 12:17:43 e2e.go:309: Error running up: exit status 2
2016/02/12 12:17:43 e2e.go:305: Step 'up' finished in 2m45.793101334s
2016/02/12 12:17:43 e2e.go:141: Error starting e2e cluster. Aborting.
exit status 1
```

```console
% ls -l /tmp/tmp.zRkaD25gbX                                                     
total 36
-rw-r----- 1 mml eng 36670 Feb 12 12:17 startupscript.e2e-test-mml-master.log.20160212-121739
% tail /tmp/tmp.zRkaD25gbX/startupscript.e2e-test-mml-master.log.20160212-121739 
Feb 12 20:17:02 e2e-test-mml-master startupscript: `/srv/kubernetes.Q0QKUD/kubernetes/server/bin/kubemark' -> `/srv/salt-new/salt/kube-bins/kubemark'
Feb 12 20:17:02 e2e-test-mml-master startupscript: `/srv/kubernetes.Q0QKUD/kubernetes/server/bin/linkcheck' -> `/srv/salt-new/salt/kube-bins/linkcheck'
Feb 12 20:17:03 e2e-test-mml-master startupscript: `/srv/kubernetes.Q0QKUD/kubernetes/addons/gcr.io~google_containers~kube-registry-proxy:0.3.tar' -> `/srv/salt-new/salt/kube-addons-images/gcr.io~google_containers~kube-registry-proxy:0.3.tar'
Feb 12 20:17:03 e2e-test-mml-master startupscript: `/srv/kubernetes.Q0QKUD/kubernetes/addons/gcr.io~google_containers~pause:2.0.tar' -> `/srv/salt-new/salt/kube-addons-images/gcr.io~google_containers~pause:2.0.tar'
Feb 12 20:17:04 e2e-test-mml-master startupscript: +++ Swapping in new configs
Feb 12 20:17:04 e2e-test-mml-master startupscript: `/srv/salt-new/salt' -> `/srv/salt'
Feb 12 20:17:04 e2e-test-mml-master startupscript: `/srv/salt-new/pillar' -> `/srv/pillar'
Feb 12 20:17:04 e2e-test-mml-master startupscript: `/srv/salt-new/reactor' -> `/srv/reactor'
Feb 12 20:17:05 e2e-test-mml-master startupscript: !!!!!!! BUG BUG BUG
Feb 12 20:17:05 e2e-test-mml-master startupscript: Finished running startup script /var/run/google.startup.script
```

--- a/cluster/gce/util.sh
+++ b/cluster/gce/util.sh
@@ -745,16 +745,25 @@ function check-cluster() {
       if [[ ${elapsed} -gt ${KUBE_CLUSTER_INITIALIZATION_TIMEOUT} ]]; then
           echo -e "${color_red}Cluster failed to initialize within ${KUBE_CLUSTER_INITIALIZATION_TIMEOUT} seconds.${color_norm}" >&2
           if [[ ${KUBE_TEST_DEBUG-} =~ ^[yY]$ ]]; then
-            local tmp_log="$(mktemp)"
-            local file
-            for file in /var/log/startupscript.log /var/log/kube-apiserver.log; do
-              echo "${MASTER_NAME}:${file} contents:"
-              if gcloud compute copy-files --project "${PROJECT}" \
-                --zone "${ZONE}" "${MASTER_NAME}:${file}" "${tmp_log}"; then
-                cat "${tmp_log}"
-              fi
+            local savedir="${E2E_REPORT_DIR-}"
+            if [[ -z "${savedir}" ]]; then
+              savedir="$(mktemp -t -d k8s-e2e.XXX)"
+            fi
+            echo "Preserving master logs in ${savedir}"
+            local logdir=/var/log
+            local basename
+            for basename in startupscript kube-apiserver; do
+              # TODO(mml): Perhaps revisit how we name logs for preservation and
+              # centralize an implementation.  Options include putting basename
+              # before hostname and including a timestamp.
+              local src="${logdir}/${basename}.log"
+              local dst="${savedir}/${MASTER_NAME}-${basename}.log"
+              echo "Copying ${MASTER_NAME}:${src}"
+              gcloud compute copy-files \
+                --project "${PROJECT}" --zone "${ZONE}" \
+                "${MASTER_NAME}:${src}" "${dst}" \
+                || true
             done
-            rm -f "${tmp_log}"
           fi
           exit 2
       fi

